lib/DxilPIXPasses/DxilPixReadOutputSig.cpp

```diff
@@ -0,0 +1,128 @@
+using namespace hlsl;
+
+const char *SemanticKindName(DXIL::SemanticKind Kind) {
+  switch (Kind) {
+  case DXIL::SemanticKind::Arbitrary:
+    return "Arbitrary";
+  case DXIL::SemanticKind::VertexID:
+    return "VertexID";
+  case DXIL::SemanticKind::InstanceID:
+    return "InstanceID";
+  case DXIL::SemanticKind::Position:
+    return "Position";
+  case DXIL::SemanticKind::RenderTargetArrayIndex:
+    return "RenderTargetArrayIndex";
+  case DXIL::SemanticKind::ViewPortArrayIndex:
+    return "ViewPortArrayIndex";
+  case DXIL::SemanticKind::ClipDistance:
+    return "ClipDistance";
+  case DXIL::SemanticKind::CullDistance:
+    return "CullDistance";
+  case DXIL::SemanticKind::OutputControlPointID:
+    return "OutputControlPointID";
+  case DXIL::SemanticKind::DomainLocation:
+    return "DomainLocation";
+  case DXIL::SemanticKind::PrimitiveID:
+    return "PrimitiveID";
+  case DXIL::SemanticKind::GSInstanceID:
+    return "GSInstanceID";
+  case DXIL::SemanticKind::SampleIndex:
+    return "SampleIndex";
+  case DXIL::SemanticKind::IsFrontFace:
+    return "IsFrontFace";
+  case DXIL::SemanticKind::Coverage:
+    return "Coverage";
+  case DXIL::SemanticKind::InnerCoverage:
+    return "InnerCoverage";
+  case DXIL::SemanticKind::Target:
+    return "Target";
+  case DXIL::SemanticKind::Depth:
+    return "Depth";
+  case DXIL::SemanticKind::DepthLessEqual:
+    return "DepthLessEqual";
+  case DXIL::SemanticKind::DepthGreaterEqual:
+    return "DepthGreaterEqual";
+  case DXIL::SemanticKind::StencilRef:
+    return "StencilRef";
+  case DXIL::SemanticKind::DispatchThreadID:
+    return "DispatchThreadID";
+  case DXIL::SemanticKind::GroupID:
+    return "GroupID";
+  case DXIL::SemanticKind::GroupIndex:
+    return "GroupIndex";
+  case DXIL::SemanticKind::GroupThreadID:
+    return "GroupThreadID";
+  case DXIL::SemanticKind::TessFactor:
+    return "TessFactor";
+  case DXIL::SemanticKind::InsideTessFactor:
+    return "InsideTessFactor";
+  case DXIL::SemanticKind::ViewID:
+    return "ViewID";
+  case DXIL::SemanticKind::Barycentrics:
+    return "Barycentrics";
+  case DXIL::SemanticKind::ShadingRate:
+    return "ShadingRate";
+  case DXIL::SemanticKind::CullPrimitive:
+    return "CullPrimitive";
+  case DXIL::SemanticKind::StartVertexLocation:
+    return "StartVertexLocation";
+  case DXIL::SemanticKind::StartInstanceLocation:
+    return "StartInstanceLocation";
+  case DXIL::SemanticKind::Invalid:
+    return "Invalid";
+  }
+  return "Unknown";
+}
+class DxilPixReadOutputSig : public ModulePass {
+public:
+  static char ID;
+  explicit DxilPixReadOutputSig() : ModulePass(ID) {}
+  bool runOnModule(Module &M) override {
+    if (OSOverride != nullptr) {
+      DxilModule &DM = M.GetOrCreateDxilModule();
+      auto const &Meta = DM.GetOutputSignature();
+      auto const &Elements = Meta.GetElements();
+      for (auto const &Element : Elements) {
+        *OSOverride << "OutputSigElement:" << Element->GetName() << ":"
```

Comment:
I'm not sure where `OSOverride` comes from, how it's used, and I'm not sure why we have to run a pass just to get the row for `SV_Position` output in upstream shader.  You could read this from the output signature part from a compiled shader, for instance.  Do you have need for the rest of the signature elements here?

Reply:
PIX needs to read the sig to get the row for SV_Position somehow so it can pass it back in to the debug instrumentation pass. This way keeps all the knowledge of the format of the signature in the compiler binary. 

I have a dastardly plan to use this also for mesh shader output visualization in PIX.